app/models/database_throttle.rb

```diff
@@ -0,0 +1,92 @@
+class DatabaseThrottle < ApplicationRecord
+  self.table_name = 'throttles'
+  belongs_to :user
+  validates :user_id, presence: true, unless: :target?
+  validates :target, presence: true, unless: :user_id?
+
+  # DEPRECATED, see Throttle
+  enum throttle_type: {
+    idv_doc_auth: 1,
+    reg_unconfirmed_email: 2,
+    reg_confirmed_email: 3,
+    reset_password_email: 4,
+    idv_resolution: 5,
+    idv_send_link: 6,
+    verify_personal_key: 7,
+    verify_gpo_key: 8,
+    proof_ssn: 9,
+    proof_address: 10,
+    phone_confirmation: 11,
+  }
+
+  # Either target or user must be supplied
+  # @param [Symbol] throttle_type
+  # @param [String] target
+  # @param [User] user
+  # @return [Throttle]
+  def self.for(throttle_type:, user: nil, target: nil)
+    throttle = if user
+      find_or_create_by(user: user, throttle_type: throttle_type)
+    elsif target
+      if !target.is_a?(String)
+        raise ArgumentError, "target must be a string, but got #{target.class}"
+      end
+      find_or_create_by(target: target, throttle_type: throttle_type)
+    else
+      raise 'Throttle must have a user or a target, but neither were provided'
+    end
+
+    throttle.reset_if_expired_and_maxed
+    throttle
+  end
+
+  # @return [Integer]
+  def increment
+    return attempts if maxed?
+    update(attempts: attempts + 1, attempted_at: Time.zone.now)
+    attempts
+  end
+
+  def throttled?
+    !expired? && maxed?
+  end
+
+  def throttled_else_increment?
+    if throttled?
+      true
+    else
+      update(attempts: attempts + 1, attempted_at: Time.zone.now)
+      false
+    end
+  end
+
+  def reset
+    update(attempts: 0)
+    self
+  end
+
+  def remaining_count
+    return 0 if throttled?
```

Comment:
coverage

Reply:
Thank you! Added spec in [5e53efd](https://github.com/18F/identity-idp/pull/6324/commits/5e53efd995f41df2c1624308d4a5cdf6d1cd4fc6)